### PR TITLE
[MM-13828] Initialize tests in each package with a new temp folder with all test resources

### PR DESCRIPTION
**Pull request by sandlis**
#### Summary
As mentioned in #10132, tests currently use `config/config.json`, which can have changes done by developers. Using changes with developers config can impact how tests run.
To avoid such impact this PR creates a new temp folder with all test resources copied/symlinked for each package having tests. To use this feature, each package needs to have TestMain and use testlib.MainHelper

#### Ticket Link
#10132 

#### Checklist
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: store/sqlstore/init_test.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package sqlstore
```
Comment (lieut-data) [inline store/sqlstore/init_test.go]: I can't remember the context here, but I'm guessing it's non-trivial to just push `initStores` and `tearDownStores` into `sqlstore_test` altogether?

File: store/sqlstore/main_test.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+func TestMain(m *testing.M) {
+	mainHelper = testlib.NewMainHelper()
+	defer mainHelper.Close()
+
+	sqlstore.InitTest()
+
+	mainHelper.Main(m)
+	sqlstore.TearDownTest()
+
```
Comment (lieut-data) [inline store/sqlstore/main_test.go]: nit: spurious extra line
```suggestion
```

File: testlib/helper.go [hunk 3]
```diff
@@ -48,20 +50,41 @@ func NewMainHelper() *MainHelper {
...
+
+	err = os.Chdir(h.TestResourcePath)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to set current working directory to %s: %s", h.TestResourcePath, err.Error()))
+	}
+
+	defer func() {
+		err := os.Chdir(prevDir)
+		if err != nil {
+			mlog.Error(fmt.Sprintf("Failed to set current working directory to %s: %s", prevDir, err.Error()))
```
Comment (lieut-data) [inline testlib/helper.go]: Propose clarifying that we're trying to revert the wd here. Also, I think it still makes sense to panic, since something is clearly very wrong.
```suggestion
			panic(fmt.Sprintf("Failed to restore current working directory to %s: %s", prevDir, err.Error()))
```

File: testlib/testResources.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
```
Comment (lieut-data) [inline testlib/testResources.go]: nit: I feel like we could name this file `resources.go` and avoid the stutter that is **test**lib.**test**Resources.go.

File: testlib/testResources.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+// See License.txt for license information.
+
+package testlib
+
+import (
+	"fmt"
```
Comment (lieut-data) [inline testlib/testResources.go]: Can you organize these imports ala what `goimports` would do?

Namely, stdlib first, then empty line, then the `mattermost-server` imports?

File: testlib/testResources.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+			testResourcesToSetup[i].src = srcPath
+		} else {
+			panic(fmt.Sprintf("Invalid resource type: %d", testResource.resType))
+		}
+	}
+}
+
+func SetupTestResources() string {
+	tempDir, err := ioutil.TempDir("", "testlibHelper")
```
Comment (lieut-data) [inline testlib/testResources.go]: Worth just using the `testlib` prefix?

File: testlib/testResources.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+			if destDir != "." {
+				err = os.MkdirAll(destDir, os.ModePerm)
+				if err != nil {
+					panic(fmt.Sprintf("failed to make dir %s: %s", destDir, err.Error()))
+				}
+			}
+
+			err = os.Symlink(testResource.src, resourceDestInTemp)
+			if err != nil {
+				panic("failed to symlink client/plugins directory to temp: " + err.Error())
```
Comment (lieut-data) [inline testlib/testResources.go]: looks like `client/plugins` was hard-coded here: can we fix the log message?

File: testlib/testResources.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+		}
+
+	}
+
+	return tempDir
+}
+
+func CleanupTestResources(tempPath string) {
+	if tempPath != "" {
+		os.RemoveAll(tempPath)
```
Comment (lieut-data) [inline testlib/testResources.go]: I'm always a little paranoid about a `os.RemoveAll`. Is it worth inlining this into `MainHelper.Close` to ensure it's only acted on in the context of the `TestResourcePath`? Then no one can accidentally  use this function.

**Comment by sandlis** [comment]
> Nice! A few comments inline, but one other requested change: while it definitely makes sense to leverage `testlib.TestMain` for all of our packages, this has the unfortunate side effect of creating and tearing down a test database for each such package, many of which don't require same. This is mildly annoying when running unit tests for those packages (while developing), since they would otherwise completely almost instantaneously.
> 
> You might consider exposing a `NewMainHelperWithoutDatabase` to effect this.

@lieut-data Thanks for the detailed review!
I was going to discuss this Database setup and teardown and some other supporting changes after getting my approach validated. I was thinking of accepting a flag in `NewMainHelper` to create/not create a database. This would avoid having multiple versions of this function if we add new resources in future to be set up before tests. Just adding flags for creating/not creating those resources would be needed.

So drawback for this approach would be having 0 values set for members of `MainHelper` which hold data relevant to those resources which won't be set up, like in this case all values inside `MainHelper` except `TestResourcePath` would be 0 values.

Please let me know whether I am thinking right.

File: build/Jenkinsfile.pr [hunk 9]
```diff
@@ -81,7 +81,6 @@ pipeline {
                     ansiColor('xterm') {
                         sh """
                             cd /go/src/github.com/mattermost/mattermost-server
-                            mv config/default.json config/config.json || echo ""
```
[the remark this answers is not in the document]
Comment (sandlis) [reply build/Jenkinsfile.pr]: This was done for making Tests work in CI. My code copies config/default.json to temp.
This was breaking tests since config/default.json was not present during test setup.

Do you want me to change the code to look for config/default.json first if it's not there then look for config/config.json?

File: store/sqlstore/init_test.go (hunk 1 above)
Comment (sandlis) [reply store/sqlstore/init_test.go]: Actually `store/sqlstore` is being imported in `testlib.helper` and `testlib.helper` is being imported in `store/sqlstore_test`, so importing `store/sqlstore` into `store/sqlstore_test` would cause circular dependency.
If you mean moving those funtions to `store/sqlstore_test` altogether then there would still be issue of circular dependency since `initStores` is using `store/storetest` and same is being imported by `testlib.helper`.

**Comment by lieut-data** [comment]
@sandlis, I agree that not having those resources initialized might surprise someone looking to leverage them in a future, database-dependent script. What if we skipped exporting them on the struct and wrote accessor methods that panicked with a more obvious message ("MainHelper not initialized with database access.")?

**Comment by sandlis** [comment]
> @sandlis, I agree that not having those resources initialized might surprise someone looking to leverage them in a future, database-dependent script. What if we skipped exporting them on the struct and wrote accessor methods that panicked with a more obvious message ("MainHelper not initialized with database access.")?

@lieut-data That's a great suggestion! It actually compliments what I was thinking for making fields of MainHelper accessible to packages which are being imported by testlib, since direct imports would cause a circular dependency.
My approach is having fields of MainHelper accessible by functions like you suggested. Packages will have Interfaces with method signatures same as functions that expose those fields. We can change the existing code as well with this approach so that anyone can refer to it in future and keep it consistent across the codebase. What do you think about this?

Also, I have added replied to some of your comments, please share your thoughts for them as well whenever you get a chance to check them.

File: build/Jenkinsfile.pr (hunk 9 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply build/Jenkinsfile.pr]: Ah, interesting -- thanks for the clear explanation. We do need to preserve the effect of this `mv` in CI for the final tar bundle delivered to end users (no `default.json` and `config.json` initialized from it).

With my recent config refactoring, we no longer even rely on `default.json` to initialize the config -- if it's missing, we just use `config.SetDefaults()` and write out `config.json` accordingly.

Perhaps we could adopt the same approach here? Instead of copying `default.json` to `config.json`, just special case that file and write out the result of serializing `config.SetDefaults` to disk?

File: store/sqlstore/init_test.go (hunk 1 above)
Comment (lieut-data) [reply store/sqlstore/init_test.go]: Ah, right. Fine as-is. Would love to decouple the sqlstore in the future to allow more flexibility.

**Comment by lieut-data** [comment]
@sandlis, I like the idea of breaking the dependency issue by defining local interfaces that `MainHelper` would satisfy. But won't each package still need a concrete instance of `MainHelper` anyway, requiring an import of `testlib`? It's likely I don't fully understand your suggestion :)

Responded to your other comments -- this is looking pretty solid!

**Comment by sandlis** [comment]
> @sandlis, I like the idea of breaking the dependency issue by defining local interfaces that `MainHelper` would satisfy. But won't each package still need a concrete instance of `MainHelper` anyway, requiring an import of `testlib`? It's likely I don't fully understand your suggestion :)
> 
> Responded to your other comments -- this is looking pretty solid!

@lieut-data I can explain it better with example :)
Let's consider MainHelper.Settings, which is actually model.SqlSettings.
MainHelper would have a method defined with signature
`func (m MainHelper) GetSqlSettings() (model.SqlSettings, error)`
So any of the package can define an interface with same signature like:
```
type TestSettings interface {
    func GetSqlSettings() (model.SqlSettings, error)
}
```
Instance of this interface would get initialized to instance of MainHelper.
Same way we'll implement methods for all the other types.
If some type is being defined by testlib itself then we'll have those types maybe moved to a sub-package of testlib which has little or no dependency on any other package.
Let me know what do you think.

**Comment by lieut-data** [comment]
Got it! Assuming I'm still following along correctly, I don't think that pattern will help us avoid a dependency problem.

To make this concrete, let's say the `app` package defines its own `TestSettings` interface. It still needs to get a concrete instance of the type, so it imports `testlib` and invokes `NewMainHelper()`. Moving parts of `testlib` to another package doesn't change anything, since `testlib` will need to import /those/ packages in order to implement the `MainHelper`: the dependency chain is now just one package longer, but would still expose any original circular problem.

This is essentially dependency injection. If the `app` package didn't need to import `testlib`, but could be given (injected) an implementation that satisfied its own `TestSettings` interface, the dependency chain would indeed be broken. But `TestMain` is the starting point of the testing, and there's no way to do dependency injection at this level.

What we really want to do is break the dependency of `testlib` on `github.com/mattermost/mattermost-server/store` and `github.com/mattermost/mattermost-server/store/sqlstore`. If we could pull up the parts of those packages that `testlib` relies on (mostly the `SqlSettings` initialization, we could probably push down the initialization of the sql store to the packages under test (e.g. back to `app`, just using `testlib` to do the database initialization but not actually /construct/ the store). Then we could safely use `testlib` even from the `sqlstore` package. This is a non-trivial change, though, and I'd propose deferring it at least from this PR's standpoint.

**Comment by sandlis** [comment]
Actually, I had some different things in mind but it still won't solve the dependency problem as cleanly as you suggested. So I'll finish the rest of the work to close this PR. We can approach the dependency problem separately.

**Comment by sandlis** [comment]
@lieut-data I am done with all the changes. Tests for enterprise is failing because the signature of testlib.NewMainHelper has changed, it now needs a bool to figure out whether db has to be set up or not. And all the values of MainHelper are private now, those needs to be accessed using accessor methods. Enterprise code is not public, right? Can someone do those changes to complete this PR?
Please review and let me know whether changes look good.

File: testlib/resources.go [hunk 10]
```diff
@@ -115,8 +121,23 @@ func SetupTestResources() string {
...
 
-func CleanupTestResources(tempPath string) {
-	if tempPath != "" {
-		os.RemoveAll(tempPath)
+func setupConfig(configDir string) {
+	var err error
+
+	err = os.Mkdir(configDir, 0700)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create config directory %s: %s", configDir, err.Error()))
```
Comment (lieut-data) [inline testlib/resources.go]: Since this is a helper method, I propose returning an `error` and panicking in `SetupTestResources` instead (or even earlier in the call chain, if you keep returning `error`). This avoids scattering panics throughout the code, and generally allows reuse of methods in other contexts without forcing panic semantics.

Comment (lieut-data) [reply testlib/resources.go]: (It also makes it easier to use `errors.Wrap` instead of having to `fmt.Sprintf` everywhere :))

File: Makefile [hunk 11]
```diff
@@ -399,7 +399,7 @@ test-compile:
 		$(GO) test $(GOFLAGS) -c $$package; \
 	done
 
-test-server: start-docker go-junit-report do-cover-file ## Runs tests.
```
[the remark this answers is not in the document]
Comment (sandlis) [reply Makefile]: Sorry, my bad. Changed it for my convenience.
Will remove it

File: app/app_test.go [hunk 12]
```diff
@@ -45,9 +45,7 @@ func TestDoAdvancedPermissionsMigration(t *testing.T) {
 	th := Setup(t)
 	defer th.TearDown()
 
-	if mainHelper.SqlSupplier == nil {
-		t.Skip("This test requires a TestStore to be run.")
-	}
+	_ = mainHelper.GetSqlSupplier()
```
[the remark this answers is not in the document]
Comment (sandlis) [reply app/app_test.go]: I thought same of not having this here and code downstream using GetSqlSupplier would panic, but I thought maybe the previous code was for stopping early if something required is not available. Should I remove it?

File: app/helper_test.go [hunk 13]
```diff
@@ -34,7 +34,8 @@ type TestHelper struct {
 }
 
 func setupTestHelper(enterprise bool, tb testing.TB) *TestHelper {
-	mainHelper.Store.DropAllTables()
+	store := mainHelper.GetStore()
+	store.DropAllTables()
```
[the remark this answers is not in the document]
Comment (sandlis) [reply app/helper_test.go]: Yes, you are right, I should have chained the calls. Actually, my initial implementation was returning `(*value, error)` and panic at places where accessor functions were called, but later I refactored it to the current implementation where I panic instead from accessor function itself to not scatter them all over the place. I missed chaining the calls at places where it is being used only once. Will change this.

File: testlib/helper.go [hunk 14]
```diff
@@ -18,16 +18,17 @@ import (
 )
 
 type MainHelper struct {
-	Settings         *model.SqlSettings
-	Store            store.Store
-	SqlSupplier      *sqlstore.SqlSupplier
-	ClusterInterface *FakeClusterInterface
+	settings         *model.SqlSettings
+	testStore        *TestStore
```
[the remark this answers is not in the document]
Comment (sandlis) [reply testlib/helper.go]: I know interfaces can be nil, but I was facing some issue at some place, I don't actually remember where actually :) Now it is working fine. Maybe some code refactoring fixed it.

File: config/file_test.go [hunk 15]
```diff
@@ -214,17 +214,17 @@ func TestFileStoreNew(t *testing.T) {
 	t.Run("relative path, file does not exist", func(t *testing.T) {
 		os.Clearenv()
 
-		err := os.MkdirAll("TestFileStoreNew/a/b/c", 0700)
+		err := os.MkdirAll("config/TestFileStoreNew/a/b/c", 0700)
```
[the remark this answers is not in the document]
Comment (sandlis) [reply config/file_test.go]: Should call to `NewMainHelper` default to setting up everything?

File: app/app_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply app/app_test.go]: So `t.Skip()` would have terminated the test immediately (without failing). I'm guessing that was to allow us to run tests without a concrete store and then without failing this test. But I'm not actually sure we ever support running these tests without a store. I'd be in favour of ripping this all out.

File: config/file_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply config/file_test.go]: Yeah, let's preserve the existing behaviour just until we can get enterprise over the line.

[the remark this answers is not in the document]
Comment (sandlis) [reply config/file_test.go]: @lieut-data `EnableResources` should be true only when we`re creating an App?
Actually while setting up resources, apart from copying config we are copying/symlinking other resources as well. Since tests run with different cwd which if don't have required resources would cause tests to fail.

[the remark this answers is not in the document]
Comment (sandlis) [reply config/file_test.go]: Actually this change would still be needed even if we do not setup default config in temp since call to `config.NewFileStore(<path>, <bool>)` resolves to `config/<path>` if config at path does not exists.

[the remark this answers is not in the document]
Comment (lieut-data) [reply config/file_test.go]: I /think/ all of those resources are only used for `App` (except maybe the weird .json edge case you had to handle, but I feel like that test pattern is invalid.) So presumably we only need resources setup then?

With respect to this particular test, the test passes today without rooting in `config/`, right? Why would it start failing with your changes?

[the remark this answers is not in the document]
Comment (sandlis) [reply config/file_test.go]: I need to check usage of test resources in packages other that packages creating an App.

This particular test is failing with my changes since it now runs with pwd set to parent of config folder while without my changes they run from config/, which is path of the package whos tests are being run. When this test runs with pwd set to config/, everything falls in place and it works.

[the remark this answers is not in the document]
Comment (lieut-data) [reply config/file_test.go]: Ok, so I think this boils back to the need to disable resources for this test: when PWD is switched to the temporary folder, but we disable resource copy and don't pre-create a `config` folder beneath it, it should no longer fall back to looking for a `config/` path, right?

[the remark this answers is not in the document]
Comment (sandlis) [reply config/file_test.go]: Yes, thats right. It would work same as it worked previously without my changes

[the remark this answers is not in the document]
Comment (sandlis) [reply config/file_test.go]: Ok so lets disable resource setup for this test. What should be criteria for disabling resource setup for other packages, should it be not creating app and not using config files?

[the remark this answers is not in the document]
Comment (lieut-data) [reply config/file_test.go]: Yeah, that would be my gut instinct -- in fact, probably any test also not setting up the store :)

**Comment by sandlis** [comment]
@lieut-data I am done with all the requested changes. I have done some additional changes apart from the ones you requested to fix some failing tests.
[This](https://github.com/mattermost/mattermost-server/pull/10261/commits/14d402005e45d3c83ac1dcd7f41527788c8f956e#diff-0c84d345c879b8d62b0c1bc73f6b22e6R86) and [this](https://github.com/mattermost/mattermost-server/pull/10261/commits/14d402005e45d3c83ac1dcd7f41527788c8f956e#diff-d74c326c126d2e2d5507081477ab8a5dR27) specially requires your attention.

Please review and let me know whether these changes look good and you want me to squash all the commits before merging.

**Comment by lieut-data** [comment]
Thanks, @sandlis! I'll get this reviewed ASAP tomorrow. No need to squash, since we do that as part of the mege here.

File: app/plugin_hooks_test.go [hunk 16]
```diff
@@ -24,7 +24,8 @@ import (
 )
 
 func compileGo(t *testing.T, sourceCode, outputPath string) {
-	dir, err := ioutil.TempDir(".", "")
+	// Using ./pkg which is symlink to this package path
```
Comment (lieut-data) [inline app/plugin_hooks_test.go]: Can you elaborate on why this is necessary?

File: app/helper_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply app/helper_test.go]: Worth pushing this through all the way? (I see it done for `GetClusterInterface`, but not yet for `GetStore`). 0/5 on this for now.

File: testlib/helper.go [hunk 17]
```diff
@@ -40,35 +48,118 @@ func NewMainHelper() *MainHelper {
...
+}
+
+func NewMainHelper() *MainHelper {
+	mainHelper := newMainHelper()
+	mainHelper.setupStore()
+	return mainHelper
+}
+
+func NewMainHelperWithOptions(options *HelperOptions) *MainHelper {
+	mainHelper := newMainHelper()
```
Comment (lieut-data) [inline testlib/helper.go]: nit: Having three-ish constructor functions seems complex. Could we just have all the logic in `NewMainHelperWithOptions`, and have `NewMainHelper()` invoke `NewMainHelperWithOptions(nil)` instead, eliminating `newMainHelper`?

File: testlib/resources.go [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+
+			testResourcesToSetup[i].src = srcPath
+		} else {
+			panic(fmt.Sprintf("Invalid resource type: %d", testResource.resType))
+		}
+	}
+}
+
+func SetupTestResources() (string, error) {
+	resolveResourcesPath()
```
Comment (lieut-data) [inline testlib/resources.go]: Can we avoid working with globals? Seems like `testResourcesToSetup` could just be declared inside `resolveResourcesPath ` and returned instead of being global in this context.

File: testlib/resources.go [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+	var config model.Config
+
+	config.SetDefaults()
+
+	err = os.Mkdir(configDir, 0700)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create config directory %s", configDir)
+	}
+
+	defaultJson := path.Join(configDir, "default.json")
```
Comment (lieut-data) [inline testlib/resources.go]: Separately, I'm making some other improvements that should render the need to copy/generate `default.json` unnecessary altogether. I'll circle back here once those are done.

File: app/plugin_hooks_test.go (hunk 16 above)
Comment (sandlis) [reply app/plugin_hooks_test.go]: [This](https://github.com/mattermost/mattermost-server/blob/14d402005e45d3c83ac1dcd7f41527788c8f956e/app/plugin_hooks_test.go#L878) code requires a non-standard package called `github.com/pkg/errors`, which is present in the vendor folder. To be able to use that package, build needs to run from within mattermost-server path.

Another option would be instead of using a non-standard package, we can use standard `errors` package and not use any of the non-standard packages except packages defined in mattermost-server.

File: app/helper_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sandlis) [reply app/helper_test.go]: [This](https://github.com/mattermost/mattermost-server/pull/10261/files/7bc274c359db126efe60ff673c8ee9ddd654f5a1..06c541d9ed183cc3244554085f292010c3c834fb#diff-dbc9c0c82ef0de5571d76d2f47f554a0R56) line in the same function uses that variable. I had checked it at all the places and double checked it now. It is assigned to a variable only when it is needed at multiple places in the same function.
I need 10/5 now :grimacing:

File: testlib/helper.go (hunk 17 above)
Comment (sandlis) [reply testlib/helper.go]: I did this way to make it easy to accommodate changes in NewMainHelperWithOptions without impacting enterprise code, like how it was failing without having the behaviour of NewMainHelper restored to the previous one. No problem changing it the way you suggested.
Does enterprise code use db or store or any other public members of MainHelper?
If yes then I think we need to set `EnableStore` to true.
Please let me know what do you think.

File: app/plugin_hooks_test.go (hunk 16 above)
Comment (hanzei) [reply app/plugin_hooks_test.go]: 3/5 on using the standard `errors` package there.

Comment (lieut-data) [reply app/plugin_hooks_test.go]: Agree with @hanzei here, that we should probably simplify the code just to use `errors` and simplify the code. This will all go away once we adopt Go modules, anyway :)

File: testlib/helper.go (hunk 17 above)
Comment (lieut-data) [reply testlib/helper.go]: I think we can accomplish what you describe above with just the two constructor functions, by writing `NewMainHelper` in terms of the `NewMainHelperWithOptions`:
```go
func NewMainHelper() *MainHelper {
        return NewMainHelperWithOptions(&HelperOptions{
            EnableStore: true,
            EnableResources: true,
        })
}
```

Yes, the enterprise code requires the store, and ideally we enable the resources automatically there as well.

**Comment by sandlis** [comment]
@lieut-data done with changes that you requested. Please review and let me know if any changes are required.

**Comment by lieut-data** [comment]
Looks great to me, @sandlis! Thanks for pushing through to completion :)

@hanzei, did you want to review this and we can get it merged? Otherwise, we can wait for @jwilander.

**Comment by sandlis** [comment]
Thanks @lieut-data It was nice working with you 
Learnt something new throughout the process :smile:

**Comment by hanzei** [comment]
I don't have the time to look into this in the next two days. I will pass this to @jwilander. :wink:
